src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -37,58 +38,88 @@
 
 namespace realm::sync {
 
-void ClientHistory::set_client_file_ident_in_wt(version_type current_version, SaltedFileIdent client_file_ident)
+void ClientHistory::set_client_reset_adjustments(
+    util::Logger& logger, version_type current_version, SaltedFileIdent client_file_ident,
+    SaltedVersion server_version, const std::vector<_impl::client_reset::RecoveredChange>& recovered_changesets)
 {
     ensure_updated(current_version); // Throws
     prepare_for_write();             // Throws
 
+    version_type client_version = m_sync_history_base_version + sync_history_size();
+    REALM_ASSERT(client_version == current_version); // For now
     Array& root = m_arrays->root;
     m_group->set_sync_file_id(client_file_ident.ident); // Throws
-    root.set(s_client_file_ident_salt_iip,
-             RefOrTagged::make_tagged(client_file_ident.salt)); // Throws
-}
 
+    size_t uploadable_bytes = 0;
+    if (recovered_changesets.empty()) {
+        // Either we had nothing to upload or we're discarding the unsynced changes
+        logger.debug("Client reset adjustments: discarding %1 history entries", sync_history_size());
+        do_trim_sync_history(sync_history_size()); // Throws
+    }
+    else {
+        // Discard all sync history before the first recovered changeset. This is
+        // required because we are going to discard our progress information and
+        // so won't know which history entries have been uploaded already.
+        auto first_version = recovered_changesets.front().version;
+        REALM_ASSERT(first_version >= m_sync_history_base_version);
+        auto discard_count = std::size_t(first_version - m_sync_history_base_version);
+        do_trim_sync_history(discard_count);
+
+        if (logger.would_log(util::Logger::Level::debug)) {
+            logger.debug("Client reset adjustments: trimming %1 history entries and updating %2 of %3 remaining "
+                         "history entries:",
+                         discard_count, recovered_changesets.size(), sync_history_size());
+            for (size_t i = 0, size = m_arrays->changesets.size(); i < size; ++i) {
+                logger.debug("- %1: ident(%2) changeset_size(%3) remote_version(%4)", i,
+                             m_arrays->origin_file_idents.get(i), m_arrays->changesets.get(i).size(),
+                             m_arrays->remote_versions.get(i));
+            }
+        }
 
-void ClientHistory::set_client_reset_adjustments(version_type current_version, SaltedFileIdent client_file_ident,
-                                                 SaltedVersion server_version, BinaryData uploadable_changeset)
-{
-    ensure_updated(current_version); // Throws
-    prepare_for_write();             // Throws
+        util::compression::CompressMemoryArena arena;
+        util::AppendBuffer<char> compressed;
+        for (auto& [changeset, version] : recovered_changesets) {
```

Comment:
I think for each recovered changeset we need to purge its reciprocal transform (if it exists) since they will be part of the merge window.

Reply:
Good catch, there definitely needs to be something done with reciprocal transforms. I'm not sure discarding them is correct. I may just be misunderstanding what reciprocal transforms are, but would it make sense for `ClientHistory::get_local_changes()` to return them instead of the original changeset when they're present?

Reply:
You can have a read on reciprocal history [here](https://github.com/10gen/baas/blob/master/realm/docs/3_ot.md#16-reciprocal-history). I am not sure using the reciprocal changes is correct either (at the time a client reset happens there may still be server changes that could have updated them, but once the client reset is done those changes are integrated without having had any impact on the reciprocal changes). This may be the reason why this approach won't work (although I am not fully convinced yet this was not an issue before too). I will give it more thought.

Reply:
After poking at the actual results it gives I'm pretty confident that recovering the reciprocal history is better than recovering the original changesets. Obviously this is all a best-effort thing that can never be fully correct, but using the reciprocal history puts the rebase origin point at the last integrated DOWNLOAD message rather than the DOWNLOAD message received before the write was originally made.

Reply:
indeed, that's what I was trying to explain above. it would be nice to add those tests.